### PR TITLE
Add missing Twig dependency and drop Twig < 2.4.0

### PR DESCRIPTION
**Pull request by greg0ire**
TODO:

- [x] drop support for Twig < 2.4.0
- [x] rename `TwigTemplate14`
- [x] add upgrade node
- [x] add changelog

## Changelog

```markdown
### Removed
- support for Twig 1
### Changed
- `Sonata\CacheBundle\Twig\TwigTemplate14` was renamed to `Sonata\CacheBundle\Twig\TwigTemplate`
```

### CODE REVIEW

**Comment by greg0ire** [comment]
🤔 maybe it should be considered optional after all... I mean people could use cache on a project that does not have Twig, right?

**Comment by Soullivaneuh** [comment]
I don't know very well this bundle.

The question is: Could the bundle break baldly if twig is not installed?

**Comment by greg0ire** [comment]
> The question is: Could the bundle break baldly if twig is not installed?

You mean it would loose its hair?

Kidding, I think we would get a lot of bug reports if it was the case. Plus, I now think that it would be best to add twig 2 support *before* adding this to require. I think I am going to move it to `require-dev`, and try to write minimal tests for the existing code.

**Comment by greg0ire** [comment]
Trying to determine what this is about when...

> Twig templates are compiled into PHP classes before using them to render contents. This option defines the base class from which all the template classes extend. Using a custom base template is discouraged because it will make your application harder to maintain

From https://symfony.com/doc/current/reference/configuration/twig.html#base-template-class

File: composer.json [hunk 1]
```diff
@@ -49,6 +50,7 @@
         ]
     },
     "autoload-dev": {
+        "psr-0": { "Twig_": "vendor/twig/twig/lib/" },
```
Comment (greg0ire) [inline composer.json]: Twig 1.3 has no autoload... should I drop support for it? This is ugly...

Comment (greg0ire) [reply composer.json]: Wait I can't, because the dependency is not required, so Composer cannot protect people who want to upgrade... next major.

Comment (core23) [reply composer.json]: What about using the `conflict` section for this?

Comment (core23) [reply composer.json]: Btw. Twig 1.3 was released looooong time ago https://github.com/twigphp/Twig/releases?after=v1.6.2

Comment (greg0ire) [reply composer.json]: The dependency is optional, so that would mean forcing people to upgrade twig even if they do not use that feature we propose... but indeed, that's a long time so maybe just drop it indeed?
Another thing to consider: we might release cachebundle 3 today. If we do, I will drop 1.3 (and 1.x) on master, as well as rename `TwigTemplate14` to just `TwigTemplate`, effectively grouping the BC-breaks together. That means the result will be the same, except that the last 2.x release will not break anything.

Comment (core23) [reply composer.json]: > The dependency is optional, so that would mean forcing people to upgrade twig even if they do not use that feature we propose... 

That's a general problem with optional deps. We can guarantee if the user is using an optional feature and needs a specific versions or not. I would risk adding a conflict section and live with a higher version instead of having a dependency version problem.

Comment (greg0ire) [reply composer.json]: Even better idea: let's do all of this in `master` instead directly. No need to risk anything on the last version of the stable branch. People that want to use twig 2 can already do so: the `TwigTemplate14` class is compatible.

File: composer.json [hunk 2]
```diff
@@ -31,7 +31,8 @@
         "php-mock/php-mock": "^1.0",
         "predis/predis": "^0.8 || ^1.0",
         "sllh/php-cs-fixer-styleci-bridge": "^2.0",
-        "symfony/phpunit-bridge": "^3.3"
+        "symfony/phpunit-bridge": "^3.3",
+        "twig/twig": "^1.3"
```
Comment (Soullivaneuh) [inline composer.json]: Should be v2 here, or I misunderstood your PR?

File: composer.json (hunk 1 above)
Comment (Soullivaneuh) [reply composer.json]: Guess what? You now have a wonderful tool to check package version since we support PHP 7 only: https://packagist.org/packages/ocramius/package-versions

With love.

Comment (Soullivaneuh) [reply composer.json]: Guess what? You now have a wonderful tool to check package version since we support PHP 7 only: https://packagist.org/packages/ocramius/package-versions

With love.

File: composer.json (hunk 2 above)
Comment (greg0ire) [reply composer.json]: Sorry, my mistake!

File: Resources/doc/reference/contextual_cache.rst [hunk 3]
```diff
@@ -10,7 +10,6 @@ You can configure the ``Twig Recorder`` with the following configuration:
 .. code-block:: yaml
 
     twig:
-        base_template_class: Sonata\CacheBundle\Twig\TwigTemplate13
-        base_template_class: Sonata\CacheBundle\Twig\TwigTemplate14
+        base_template_class: Sonata\CacheBundle\Twig\TwigTemplate
 
-// todo add more information
\ No newline at end of file
+// todo add more information
```
Comment (Soullivaneuh) [inline Resources/doc/reference/contextual_cache.rst]: Didn't get this change. Oo

File: Tests/Twig/TwigTemplateTest.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\CacheBundle\Tests\Twig;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\CacheBundle\Twig\TwigTemplate;
+
+class TwigTemplateTest extends TestCase
```
Comment (Soullivaneuh) [inline Tests/Twig/TwigTemplateTest.php]: `final`

File: Resources/doc/reference/contextual_cache.rst (hunk 3 above)
Comment (greg0ire) [reply Resources/doc/reference/contextual_cache.rst]: It's just a newline

File: composer.json [hunk 5]
```diff
@@ -31,7 +31,8 @@
         "php-mock/php-mock": "^1.0",
         "predis/predis": "^0.8 || ^1.0",
         "sllh/php-cs-fixer-styleci-bridge": "^2.0",
-        "symfony/phpunit-bridge": "^3.3"
+        "symfony/phpunit-bridge": "^3.3",
+        "twig/twig": "~2.4.0"
```
Comment (Soullivaneuh) [inline composer.json]: `^2.4`

**Comment by jordisala1991** [comment]
It should be: Template::ANY_CALL

https://github.com/twigphp/Twig/blob/v2.4.4/lib/Twig/Template.php#L26

TemplateInterface is deprecated since... 2012

https://github.com/twigphp/Twig/blob/83c7fc643278acdc249fbdc17ff5c80be73b5d7c/doc/deprecated.rst

**Comment by greg0ire** [comment]
@jordisala1991 I checked the parent... `getAttribute` does not seem to exist anymore! :(

**Comment by greg0ire** [comment]
Since https://github.com/twigphp/Twig/commit/640568437f673a5f18ea6878c0f93d22531effde

**Comment by jordisala1991** [comment]
Here it is that code removal:

https://github.com/twigphp/Twig/commit/640568437f673a5f18ea6878c0f93d22531effde

**Comment by jordisala1991** [comment]
Heh, you're faster

**Comment by greg0ire** [comment]
Too slow :P

**Comment by greg0ire** [comment]
You're faster at saying I'm faster :P

**Comment by greg0ire** [comment]
Not sure what to do about this anymore ...

**Comment by jordisala1991** [comment]
So... does it make sense this whole twig thing?

**Comment by greg0ire** [comment]
> So... does it make sense this whole twig thing?

This PR makes no sense. And the functionality... I don't know what it is used for in the first place. I know it has to do with page and block. Maybe @rvanginneken or @acrobat can help?

**Comment by greg0ire** [comment]
<details>

```
rg recorder [Ss]onata*
SonataCacheBundle/SonataCacheBundle.php
36:            call_user_func([$baseTemplateClass, 'attachRecorder'], $this->container->get('sonata.cache.recorder'));

sonata-cache/src/CacheManager.php
32:    protected $recorder;
93:    public function setRecorder(Recorder $recorder): void
95:        $this->recorder = $recorder;
103:        return $this->recorder;

sonata-cache/src/CacheManagerInterface.php
59:     * Sets the recorder.
61:     * @param Recorder $recorder
63:    public function setRecorder(Recorder $recorder): void;
66:     * Gets the recorder.

SonataCacheBundle/DependencyInjection/SonataCacheExtension.php
81:        $recorder = $container->getDefinition('sonata.cache.model_identifier');
83:            $recorder->addMethodCall('addClass', [$class, $method]);
86:        $cacheManager->addMethodCall('setRecorder', [new Reference($config['cache_invalidation']['recorder'])]);

SonataCacheBundle/Twig/TwigTemplate.php
22:    protected static $recorder;
27:     * @param \Sonata\Cache\Invalidation\Recorder $recorder
29:    public static function attachRecorder(Recorder $recorder)
31:        self::$recorder = $recorder;
48:        if (self::$recorder && is_object($object)) {
49:            self::$recorder->add($object);

SonataCacheBundle/DependencyInjection/Configuration.php
45:                    ->scalarNode('recorder')->defaultValue('sonata.cache.recorder')->end()

SonataPageBundle/DependencyInjection/Configuration.php
125:                    ->scalarNode('recorder')->defaultValue('sonata.cache.recorder')->end()

SonataPageBundle/Cache/BlockEsiCache.php
52:    protected $recorder;
62:     * @param Recorder                     $recorder         The cache recorder to build the contextual key
64:    public function __construct($token, array $servers, RouterInterface $router, $purgeInstruction, BlockRendererInterface $blockRenderer, BlockContextManagerInterface $contextManager, array $managers = [], Recorder $recorder = null)
71:        $this->recorder = $recorder;
121:        if ($this->recorder) {
122:            $this->recorder->add($blockContext->getBlock());
123:            $this->recorder->push();
128:        if ($this->recorder) {
129:            $keys = $this->recorder->pop();

sonata-cache/tests/Invalidation/RecorderTest.php
52:        $recorder = new Recorder($collection);
54:        $recorder->push();
56:        $recorder->add($m1);
58:        $recorder->push();
59:        $recorder->add($m1);
60:        $recorder->add($m2);
61:        $recorder->add($m2);
63:        $innerKeys = $recorder->pop();
64:        $recorder->add($m3);
66:        $keys = $recorder->pop();

SonataBlockBundle/Templating/Helper/BlockHelper.php
268:            $recorder = null;
270:                $recorder = $this->cacheManager->getRecorder();
272:                if ($recorder) {
273:                    $recorder->add($blockContext->getBlock());
274:                    $recorder->push();
279:            $contextualKeys = $recorder ? $recorder->pop() : [];

SonataCacheBundle/Resources/config/cache.xml
8:        <service id="sonata.cache.recorder" class="Sonata\Cache\Invalidation\Recorder">

SonataPageBundle/Resources/config/cache.xml
17:            <argument type="service" id="sonata.cache.recorder"/>

SonataPageBundle/Tests/DependencyInjection/ConfigurationTest.php
144:                'recorder' => 'sonata.cache.recorder',
229:                'recorder' => 'sonata.cache.recorder',

SonataPageBundle/Resources/doc/reference/advanced_configuration.rst
55:                recorder:             sonata.cache.recorder

sonata-sandbox/app/config/sonata/sonata_page.yml
21:        recorder: sonata.page.cache.recorder

sonata-project.org/app/config/sonata/sonata_page.yml
31:        recorder: sonata.page.cache.recorder

```

</details>

**Comment by greg0ire** [comment]
I think this is used to provide a way to invalidate esi blocks in the page bundle: https://github.com/sonata-project/SonataPageBundle/blob/3.x/Cache/BlockEsiCache.php#L128-L136

**Comment by rvanginneken** [comment]
@greg0ire I have no idea what the recorder's job is, but it looks like this implementation isn't working anymore. Like said earlier getAttribute is removed and not called anymore. There is a solution offered in https://github.com/twigphp/Twig/issues/2338, which I haven't looked into yet. I'll do a quick checkout of this bundle and have a look. No promises ;)

tldr: IMO PR could be closed without merging atm. Bug report for missing functionality could be created?

**Comment by greg0ire** [comment]
Yeah let's not risk an accidental merge.

**Comment by rvanginneken** [comment]
@greg0ire I would suggest putting a deprecation notice in there, so someone who is actually using it can pick it up or give a usecase. I think it's a bad idea to just "refactor" this, having no idea how exactly it's supposed to work and what the desired results are.

**Comment by greg0ire** [comment]
In this case "refactor" = "remove" :sweat_smile: But indeed, a deprecation should be in order!

**Comment by rvanginneken** [comment]
I'll make a quick PR.
